[Codegen] Bubble up Transpose attention V and try fuse with others before attention

raikonenfnu opened the pull request:
Flash Attention transpose_V variant is significantly faster than the non-transpose_V variant. This is due to many matmul intrinsics being mmtb by default. Hence, doing FA transpose_V will allow for better/more contiguous reads from shared memory to register, improving the attention performance quite a bit.

This PR exposes the attention_transposeV form by generating a linalg.transpose on the V during bubbling up of transpose S.T we can give the graph some opportunities to fuse the transpose-V to it's producer. I have also confirmed that if we do not find any producer, the transpose will indeed fuse back with the attenionOp. Hence worse case, we will get same perf as before this PR.

Additionally, we modify elementwise op fusion to try fuse transpose with other ops before letting it get fused back into attention.

raikonenfnu committed:
[Torch][Input]

Flash Attention transpose_V variant is significantly faster than the
non-transpose_V variant. This is due to many matmul intrinsics being
mmtb by default. Hence, doing FA transpose_V will allow for better/more
contiguous reads from shared memory to register, improving the attention
performance quite a bit.

This PR exposes the attention_transposeV form by default through the
torch lowering by generating a linalg.transpose on the V S.T we can give
the graph some opportunities to fuse the transpose-V to it's producer. I
have also confirmed that if we do not find any producer, the transpose
will indeed fuse back with the attenionOp. Hence worse case, we will get
same perf as before this PR.

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu committed:
Move transpose into bubbling patterns during transpose propagation

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu committed:
fix precommit

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu committed:
Add spec to get match tuned perf of transposeV variants of mmt

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu committed:
Fix CIs

1. Update compile cmd flag to ToM sdxl-script with no horizontal-matmul
   fusion, and update dispatch count numbers.
2. Fix bazel erroring out by removing non necessary check.

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu committed:
test adding back fuse horizontal for correctness

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu committed:
Modify time since we got a bit faster + scheduled unet dispatch count

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu committed:
Add control function on bubbling linalgExt for safety

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

raikonenfnu committed:
update mi250 pkgci numbers

Signed-off-by: Stanley Winata <stanley.winata@amd.com>

IanWood1 commented on .github/workflows/pkgci_regression_test.yml:
```diff
@@ -242,17 +242,17 @@ jobs:
             --goldentime-rocm-clip-ms 15.5 \
             --goldentime-rocm-vae-ms 80.0 \
-            --goldendispatch-rocm-unet 1531 \
+            --goldendispatch-rocm-unet 1602 \
```
Since you are updating fp16 & fp8 runtime goldens, could you also lower unet/clip/vae/e2e here as well? Was there an improvement there as well?

raikonenfnu replied:
didn't check but can do that :)

saienduri reviewed (approved):
Looks good from CI side

IanWood1 commented on compiler/src/iree/compiler/Dialect/LinalgExt/Transforms/TransposeFusion.cpp:
```diff
@@ -101,6 +102,103 @@ struct FuseTransposeWithAttentionOp final
+      return failure();
+    }
+    // Check that V has standard map/non transposed V.
+    AffineExpr k2Dim =
+        rewriter.getAffineDimExpr(maybeOpInfo->getK2Dims().back());
+    AffineExpr nDim = rewriter.getAffineDimExpr(maybeOpInfo->getNDims().back());
```
Do you need to check that the K2 dims and the N Dims only have a single element?

IanWood1 replied:
Whoops, the check is right above

IanWood1 commented on compiler/src/iree/compiler/DispatchCreation/ElementwiseOpFusion.cpp:
```diff
@@ -135,24 +134,35 @@ void ElementwiseOpFusionPass::runOnOperation() {
       };
-  linalg::populateElementwiseOpsFusionPatterns(fusionPatterns,
+
+  RewritePatternSet linalgFusionPatterns(context);
```
I feel like this change shouldn't require any changes to ElementwiseOpFusionPass and transpose on the input should be fused with the producer during transpose propagation. It just seems a bit weird that there is this implicit connection between the passes where one introduces the transpose and the other makes sure to fuse it in the optimal way.

That might require some serious changes to transpose propagation that probably aren't in the scope of this PR, so this makes sense for now. Also, `populateFuseLinalgExtOpsWithTransposes` is a bit hacky so it makes sense to run it after everything.

IanWood1 replied:
@qedawkins I think you were the one mainly working on that pass. Is there a reason why we don't fuse transposes with elementwise ops during that pass?

raikonenfnu replied:
Yeah agreed, ideally Transpose Propagation takes care of everything, but currently for some reason the transposeOp get sunk back down. i.e if I do `test-bubbling-only` it works as expected. That's why we need the change to `populateFuseLinalgExtOpsWithTransposes` to guarantee the fuse into producer.

raikonenfnu committed:
update unet and e2e number

unet improve from 80.5 ~ 80.9 ms to ~ 78.3 ms
e2e number improve from ~331.0 ms to ~324 ms

Signed-off-by: Stanley Winata <stanley.winata@amd.com>
